### ISSUE TITLE
ES2015 updates

### ISSUE THREAD

**Issue by jmeas** [tbranyen/diffhtml#40]
Migrating the conversation from #37 

===

tl;dr

@tbranyen was finding it difficult to get some test-related features working with ES2015. Namely, src maps, cvg reports, and browser tests.

I've spent far too much time in the past year thinking about using ES2015 with popular testing libraries. Much of that work exists in a usable form over in this repo: [generator-babel-boilerplate](https://github.com/babel/generator-babel-boilerplate)

Basically all that I'm doing is porting this lib over to use that boilerplate, or ideas from that boilerplate, with changes where necessary.

I'm opening a separate issue in case I've got more questions, and to keep it separate from that PR, which I've closed.

===

Todo:

- [ ] Add linting back before test and coverage gulp tasks
- [ ] Update package.json tasks
- [ ] Remove unused devDeps
- [ ] Split `test-browser` into one that builds app for phantomJS testin' and one for in-browser testin'

Changes

- Prollyfill no longer patches `window` to add `TransitionStateError`. This makes testing harder, most developers think it is bad manners to patch window, and it is unnecessary as `TransitionStateError` is available under `diff.TransitionStateError`

**Comment by jmeas** [comment]
Note to self for the future:

I attempted to use SimpleJSDom to make the tests pass. All of them did except for the pesky integration tests that modified the HTML or document. I thought "no worries," and configured the tests to pass in-browser. This enabled me to use PhantomJS for CLI testing and coverage. Easy, right?

Not quite – I was instrumenting the **source** files, and – d'uh – those aren't used in the tests in the spec runner.

So I'll need to come up with a new system for this. Honestly, I was pretty lazy about figuring out the JSDom problem, so it's entirely possible that it's solvable. Otherwise, I'll need to think up a clever way to instrument the src files and get those to run in PhantomJS.

In any event, these are problems for another day.

**Comment by tbranyen** [comment]
Can we skip that test for now, because it's most likely related to a different test?

**Comment by tbranyen** [comment]
That test was bad from the start, but it was a stop gap to ensure that my cleanup was working, I wasn't sure how else to test something meant to be 100% internal.

**Comment by jmeas** [comment]
> Can we skip that test for now, because it's most likely related to a different test?

Fo' sho'. I'll comment it out and keep looking into the JSDom fail.

**Comment by jmeas** [comment]
Alright, really gonna head to bed, but here's [a branch](https://github.com/jmeas/diffhtml/tree/gulp-stuff) with some progress.

What's working:

- Basic coverage
- Browser tests
- ES2015 can be used in the tests
- Source maps (probably lol, didn't extensively test yet but I use this general setup on all of my libs/apps and I never have problems)

To make it work, 3 tests had to be turned off (the mem mgmt test and the two tests that replaced the whole document) and 1 feature removed (attaching the TransitionStateError to window).

I suspect that getting JSDom to make those document replacement tests pass will be a substantial effort, but I'll put a few hours into it and see where it takes me.

To test:

1. Add my remote
2. Switch to the `gulp-stuff` branch
3. `npm i`
4 Run these commands for stuff:
  - `gulp coverage` for coverage reports. The reports will be emitted in the CLI and also to the `coverage` directory
  - `gulp test-browser` to build the lib for the browser. After it's built, open up `test/index.html` in your browser to run the tests. Webpack will intelligently reload as you make changes (or it should, I didn't test it yet for this particular lib tbqh)
  - `gulp test` for CLI tests, for, say, travis

**Comment by tbranyen** [comment]
Awesome, thanks so much @jmeas. I'm still working on my updates to the master branch. I'll rebase my stuff with yours once I'm done and test it out.

**Comment by jmeas** [comment]
Closing; superseded by #42
